// aws-cdk-lib/aws-elasticloadbalancingv2-actions/lib/cognito-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthenticateCognitoAction = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const elbv2 = require("../../aws-elasticloadbalancingv2");
/**
 * A Listener Action to authenticate with Cognito
 */
class AuthenticateCognitoAction extends elbv2.ListenerAction {
    static config(options) {
        return {
            userPoolArn: options.userPool.userPoolArn,
            userPoolClientId: options.userPoolClient.userPoolClientId,
            userPoolDomain: options.userPoolDomain.domainName,
            authenticationRequestExtraParams: options.authenticationRequestExtraParams,
            onUnauthenticatedRequest: options.onUnauthenticatedRequest,
            scope: options.scope,
            sessionCookieName: options.sessionCookieName,
            sessionTimeout: options.sessionTimeout?.toSeconds().toString(),
        };
    }
    /**
     * Authenticate using an identity provide (IdP) that is compliant with OpenID Connect (OIDC)
     */
    constructor(options) {
        super({
            type: 'authenticate-cognito',
            authenticateCognitoConfig: AuthenticateCognitoAction.config(options),
        }, options.next);
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_actions_AuthenticateCognitoActionProps(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, AuthenticateCognitoAction);
            }
            throw error;
        }
        this.addRuleAction({
            type: 'authenticate-cognito',
            authenticateCognitoConfig: {
                ...AuthenticateCognitoAction.config(options),
                sessionTimeout: options.sessionTimeout?.toSeconds(),
            },
        });
    }
}
_a = JSII_RTTI_SYMBOL_1;
AuthenticateCognitoAction[_a] = { fqn: "aws-cdk-lib.aws_elasticloadbalancingv2_actions.AuthenticateCognitoAction", version: "2.77.0" };
exports.AuthenticateCognitoAction = AuthenticateCognitoAction;
