// aws-cdk-lib/core/lib/custom-resource-provider/cross-region-export-providers/export-reader-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportReader = void 0;
const path = require("path");
const constructs_1 = require("constructs");
const types_1 = require("./types");
const cfn_resource_1 = require("../../cfn-resource");
const custom_resource_1 = require("../../custom-resource");
const lazy_1 = require("../../lazy");
const stack_1 = require("../../stack");
const custom_resource_provider_1 = require("../custom-resource-provider");
/**
 * Creates a custom resource that will return a list of stack imports from a given
 * The export can then be referenced by the export name.
 *
 * @internal - this is intentionally not exported from core
 */
class ExportReader extends constructs_1.Construct {
    static getOrCreate(scope, uniqueId, _props = {}) {
        const stack = stack_1.Stack.of(scope);
        const existing = stack.node.tryFindChild(uniqueId);
        return existing
            ? existing
            : new ExportReader(stack, uniqueId);
    }
    constructor(scope, id, _props = {}) {
        super(scope, id);
        this.importParameters = {};
        const stack = stack_1.Stack.of(this);
        const resourceType = 'Custom::CrossRegionExportReader';
        const serviceToken = custom_resource_provider_1.CustomResourceProvider.getOrCreate(this, resourceType, {
            codeDirectory: path.join(__dirname, 'cross-region-ssm-reader-handler'),
            runtime: (0, custom_resource_provider_1.builtInCustomResourceProviderNodeRuntime)(this),
            policyStatements: [{
                    Effect: 'Allow',
                    Resource: stack.formatArn({
                        service: 'ssm',
                        resource: 'parameter',
                        resourceName: `${types_1.SSM_EXPORT_PATH_PREFIX}${stack.stackName}/*`,
                    }),
                    Action: [
                        'ssm:AddTagsToResource',
                        'ssm:RemoveTagsFromResource',
                        'ssm:GetParameters',
                    ],
                }],
        });
        const properties = {
            region: stack.region,
            prefix: stack.stackName,
            imports: lazy_1.Lazy.any({ produce: () => this.importParameters }),
        };
        this.customResource = new custom_resource_1.CustomResource(this, 'Resource', {
            resourceType: resourceType,
            serviceToken,
            properties: {
                ReaderProps: properties,
            },
        });
    }
    /**
     * This is the only way to add a dependency on a custom resource currently
     */
    addDependency(resource) {
        const customResource = this.customResource.node.tryFindChild('Default');
        if (customResource && cfn_resource_1.CfnResource.isCfnResource(customResource)) {
            customResource.addDependsOn(resource);
        }
    }
    /**
     * Register a reference with the writer and returns a CloudFormation Stack export by name
     *
     * The value will be "exported" via the ExportWriter. It will perform
     * the export by creating an SSM parameter in the region that the consuming
     * stack is created.
     *
     * @param exports map of unique name associated with the export to SSM Dynamic reference
     */
    importValue(name, value) {
        this.importParameters[name] = value.toString();
        return this.customResource.getAtt(name);
    }
}
exports.ExportReader = ExportReader;
